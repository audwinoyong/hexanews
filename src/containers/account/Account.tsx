/**
 * Account page container for user to edit their profile
 */

import * as React from 'react'
import { ValidatorForm, TextValidator } from 'react-material-ui-form-validator'
import { connect } from 'react-redux'
import Button from '@material-ui/core/Button'
import Divider from '@material-ui/core/Divider'
import CircularProgress from '@material-ui/core/CircularProgress'
import { Typography } from '@material-ui/core'
import TextField from '@material-ui/core/TextField'
import { Dispatch } from 'redux'

import * as actions from './actions'
import styles from './styles'
import { User } from '../../domain/model/User'
import { accountString } from '../../constants/string'
import selector, { StateProps } from './selector'

type Props = {
  user: User,
  dispatch: Dispatch<any>,
  isEditingUser: boolean,
  isChangingPassword: boolean,
} & StateProps

interface ComponentState {
  email: string,
  name: string,
  description: string,
  currentPassword: string,
  newPassword: string,
  confirmNewPassword: string,
  submitted: boolean,
}

class Account extends React.Component<Props, ComponentState> {

  constructor(props: Props) {
    super(props)
    const { user } = this.props
    this.state = {
      email: user.email,
      name: user.name,
      description: user.description ? user.description : '',
      currentPassword: '',
      newPassword: '',
      confirmNewPassword: '',
      submitted: false,
    }
  }

  handleNameChange = (event: any) => {
    this.setState({ name: event.target.value })
  }

  handleDescriptionChange = (event: any) => {
    this.setState({ description: event.target.value })
  }

  handleCurrentPasswordChange = (event: any) => {
    this.setState({ currentPassword: event.target.value })
  }

  handleNewPasswordChange = (event: any) => {
    this.setState({ newPassword: event.target.value })
  }

  handleConfirmNewPasswordChange = (event: any) => {
    this.setState({ confirmNewPassword: event.target.value })
  }

  onSaveProfile = async () => {
    const { dispatch } = this.props
    const { email, name, description } = this.state
    const _id = localStorage.getItem('id') || ''
    const createdAt = localStorage.getItem('createdAt') || ''

    const edittedUser = { _id, email, name, description, createdAt }
    await dispatch(actions.editUser(edittedUser))
  }

  onChangePassword = async () => {
    const { dispatch } = this.props
    const { email, currentPassword, newPassword } = this.state

    await dispatch(actions.changePassword(email, currentPassword, newPassword))
  }

  public render() {
    const { isEditingUser, isChangingPassword } = this.props
    const { email, name, description, currentPassword, newPassword, confirmNewPassword, submitted } = this.state
    return (
      <div style={styles.container}>
        <Typography
          variant='display1'
          component='h1'
          color='textPrimary'
          gutterBottom
        >
          {accountString.myAccount}
        </Typography>
        <div style={styles.sectionContainer}>
          <Typography
            variant='headline'
            component='h2'
            color='textPrimary'
            gutterBottom
          >
            {accountString.editProfile}
          </Typography>
          <ValidatorForm
            ref='saveProFileForm'
            onSubmit={this.setState({ submitted: true })}
          >
            <TextField
              id='email'
              label='Email'
              value={email}
              disabled
              style={styles.textField}
              margin='normal'
            />
            <TextValidator
              label='Name'
              name='name'
              value={name}
              onChange={this.handleNameChange}
              style={styles.textField}
              margin='normal'
              validators={['required', 'isEmail']}
              errorMessages={['this field is required', 'email is not valid']}
            />
            <TextValidator
              label='Description'
              name='description'
              value={description}
              onChange={this.handleDescriptionChange}
              style={styles.textField}
              margin='normal'
              validators={['minCharacter:0', 'maxCharacter:100']}
              errorMessages={['this field is required', 'maximum 100 characters']}
            />
            {/* <TextField
              id='name'
              label='Name'
              value={name}
              onChange={this.handleNameChange}
              style={styles.textField}
              margin='normal'
            /> */}
            {/* <TextField
              id='description'
              label='Description'
              value={description}
              onChange={this.handleDescriptionChange}
              style={styles.textField}
              margin='normal'
            /> */}
            <Button
              variant='outlined'
              size='small'
              type='submit'
              disabled={submitted}
              component='button'
              style={styles.button}
              // onClick={this.onSaveProfile}
            >
              {isEditingUser ? <CircularProgress size={22} /> : accountString.saveButton}
            </Button>
          </ValidatorForm>
        </div>
        <Divider style={styles.sectionDivider} />
        <div style={styles.sectionContainer}>
          <Typography
            variant='headline'
            component='h2'
            color='textPrimary'
          >
            {accountString.changePassword}
          </Typography>
          <TextField
            id='currentPassword'
            label='Current Password'
            value={currentPassword}
            onChange={this.handleCurrentPasswordChange}
            style={styles.textField}
            margin='normal'
            type='password'
          />
          <TextField
            id='newPassword'
            label='New Password'
            value={newPassword}
            onChange={this.handleNewPasswordChange}
            style={styles.textField}
            margin='normal'
            type='password'
          />
          <TextField
            id='confirmNewPassword'
            label='Confirm New Password'
            value={confirmNewPassword}
            onChange={this.handleConfirmNewPasswordChange}
            style={styles.textField}
            margin='normal'
            type='password'
          />
        </div>
        <div>
          <Button
            variant='outlined'
            size='small'
            component='button'
            style={styles.button}
            onClick={this.onChangePassword}
          >
            {isChangingPassword ? <CircularProgress size={22} /> : accountString.changePassword}
          </Button>
        </div>
      </div>
    )
  }
}

export default connect(selector)(Account)