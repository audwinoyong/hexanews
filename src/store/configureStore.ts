/**
 * Configuration file for redux store
 */

import thunk from 'redux-thunk'
import { applyMiddleware, createStore } from 'redux'
import { createLogger } from 'redux-logger'
import { connectRouter, routerMiddleware } from 'connected-react-router'
import { composeWithDevTools } from 'remote-redux-devtools'
import { History } from 'history'

import rootReducer from '../reducers'

const isDebuggingInChrome = !!window.navigator.userAgent

/**
 * Function to create the initial store state for application
 * @param initialState the initial state for the store
 * @param history the navigation variable for the application
 */
export default function (initialState = {}, history: History) {
  return createStore(
    connectRouter(history)(rootReducer),
    initialState,
    composeWithDevTools(applyMiddleware(
      routerMiddleware(history),
      thunk,
      createLogger({
        predicate: () => isDebuggingInChrome,
        collapsed: true,
        duration: true,
      }),
    )),
  )
}