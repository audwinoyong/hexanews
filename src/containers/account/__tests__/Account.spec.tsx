import * as React from 'react'
import * as renderer from 'react-test-renderer'
import * as Enzyme from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'

import rootReducer from '../../../reducers'
import { Account } from '../Account'

const mockStore = configureStore([thunk])
Enzyme.configure({ adapter: new Adapter() })

describe('Account', () => {
  it('renders Account', () => {
    const store = mockStore(rootReducer(undefined as any, { type: '' }))
    const component = renderer.create(
      <Provider store={store}>
        <Account
        />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})