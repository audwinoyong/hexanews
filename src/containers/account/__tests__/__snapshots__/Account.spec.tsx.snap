// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account renders Account 1`] = `
<div
  className="MuiGrid-container-1"
  style={
    Object {
      "marginBottom": "4%",
      "marginTop": "4%",
    }
  }
>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
  />
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-10-39 MuiGrid-grid-md-6-63"
  >
    <h1
      className="MuiTypography-root-98 MuiTypography-display1-102 MuiTypography-colorTextPrimary-120 MuiTypography-gutterBottom-115"
    >
      My Account
    </h1>
    <div
      className=""
      style={
        Object {
          "marginTop": 24,
        }
      }
    >
      <form
        onSubmit={[Function]}
      >
        <h2
          className="MuiTypography-root-98 MuiTypography-headline-103 MuiTypography-colorTextPrimary-120 MuiTypography-gutterBottom-115"
        >
          Edit Profile
        </h2>
        <div
          aria-describedby="email-helper-text"
          className="MuiFormControl-root-123 MuiFormControl-marginDense-125"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <label
            className="MuiFormLabel-root-132 MuiInputLabel-root-127 MuiInputLabel-formControl-128 MuiInputLabel-animated-131 MuiInputLabel-marginDense-129"
            data-shrink={false}
            htmlFor="email"
          >
            Email
          </label>
          <div
            className="MuiInput-root-139 MuiInput-disabled-142 MuiInput-formControl-140 MuiInput-underline-143"
          >
            <input
              aria-invalid={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-147 MuiInput-disabled-142 MuiInput-inputMarginDense-148"
              defaultValue={undefined}
              disabled={true}
              id="email"
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={false}
              rows={undefined}
              type="text"
              value=""
            />
          </div>
          <p
            className="MuiFormHelperText-root-152 MuiFormHelperText-marginDense-155"
            id="email-helper-text"
          >
             
          </p>
        </div>
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-123 MuiFormControl-marginDense-125"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <label
            className="MuiFormLabel-root-132 MuiInputLabel-root-127 MuiInputLabel-formControl-128 MuiInputLabel-animated-131 MuiInputLabel-marginDense-129"
            data-shrink={false}
            htmlFor={undefined}
          >
            Name
          </label>
          <div
            className="MuiInput-root-139 MuiInput-formControl-140 MuiInput-underline-143"
          >
            <input
              aria-invalid={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-147 MuiInput-inputMarginDense-148"
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={false}
              rows={undefined}
              type="text"
              value=""
            />
          </div>
          <p
            className="MuiFormHelperText-root-152 MuiFormHelperText-marginDense-155"
            id={undefined}
          >
             
          </p>
        </div>
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-123 MuiFormControl-marginDense-125"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <label
            className="MuiFormLabel-root-132 MuiInputLabel-root-127 MuiInputLabel-formControl-128 MuiInputLabel-animated-131 MuiInputLabel-marginDense-129"
            data-shrink={false}
            htmlFor={undefined}
          >
            Description
          </label>
          <div
            className="MuiInput-root-139 MuiInput-formControl-140 MuiInput-underline-143"
          >
            <input
              aria-invalid={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-147 MuiInput-inputMarginDense-148"
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              name="description"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={false}
              rows={undefined}
              type="text"
              value=""
            />
          </div>
          <p
            className="MuiFormHelperText-root-152 MuiFormHelperText-marginDense-155"
            id={undefined}
          >
             
          </p>
        </div>
        <h2
          className="MuiTypography-root-98 MuiTypography-body1-107 MuiTypography-colorTextPrimary-120"
          style={
            Object {
              "color": "#E53935",
            }
          }
        >
          
        </h2>
        <button
          className="MuiButtonBase-root-185 MuiButtonBase-disabled-186 MuiButton-root-159 MuiButton-text-161 MuiButton-outlined-167 MuiButton-sizeSmall-182 MuiButton-disabled-179"
          disabled={true}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "marginTop": 10,
            }
          }
          tabIndex="-1"
          type="submit"
        >
          <span
            className="MuiButton-label-160"
          >
            Save Profile
          </span>
        </button>
      </form>
    </div>
    <hr
      className="MuiDivider-root-195"
      style={
        Object {
          "marginTop": 50,
        }
      }
    />
    <div
      className=""
      style={
        Object {
          "marginTop": 24,
        }
      }
    >
      <form
        onSubmit={[Function]}
      >
        <h2
          className="MuiTypography-root-98 MuiTypography-headline-103 MuiTypography-colorTextPrimary-120"
        >
          Change Password
        </h2>
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-123 MuiFormControl-marginDense-125"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <label
            className="MuiFormLabel-root-132 MuiInputLabel-root-127 MuiInputLabel-formControl-128 MuiInputLabel-animated-131 MuiInputLabel-marginDense-129"
            data-shrink={false}
            htmlFor={undefined}
          >
            Current Password
          </label>
          <div
            className="MuiInput-root-139 MuiInput-formControl-140 MuiInput-underline-143"
          >
            <input
              aria-invalid={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-147 MuiInput-inputType-150 MuiInput-inputMarginDense-148"
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              name="currentPassword"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={false}
              rows={undefined}
              type="password"
              value=""
            />
          </div>
          <p
            className="MuiFormHelperText-root-152 MuiFormHelperText-marginDense-155"
            id={undefined}
          >
             
          </p>
        </div>
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-123 MuiFormControl-marginDense-125"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <label
            className="MuiFormLabel-root-132 MuiInputLabel-root-127 MuiInputLabel-formControl-128 MuiInputLabel-animated-131 MuiInputLabel-marginDense-129"
            data-shrink={false}
            htmlFor={undefined}
          >
            New Password
          </label>
          <div
            className="MuiInput-root-139 MuiInput-formControl-140 MuiInput-underline-143"
          >
            <input
              aria-invalid={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-147 MuiInput-inputType-150 MuiInput-inputMarginDense-148"
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              name="newPassword"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={false}
              rows={undefined}
              type="password"
              value=""
            />
          </div>
          <p
            className="MuiFormHelperText-root-152 MuiFormHelperText-marginDense-155"
            id={undefined}
          >
             
          </p>
        </div>
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-123 MuiFormControl-marginDense-125"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <label
            className="MuiFormLabel-root-132 MuiInputLabel-root-127 MuiInputLabel-formControl-128 MuiInputLabel-animated-131 MuiInputLabel-marginDense-129"
            data-shrink={false}
            htmlFor={undefined}
          >
            Confirm New Password
          </label>
          <div
            className="MuiInput-root-139 MuiInput-formControl-140 MuiInput-underline-143"
          >
            <input
              aria-invalid={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-147 MuiInput-inputType-150 MuiInput-inputMarginDense-148"
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              name="confirmNewPassword"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={false}
              rows={undefined}
              type="password"
              value=""
            />
          </div>
          <p
            className="MuiFormHelperText-root-152 MuiFormHelperText-marginDense-155"
            id={undefined}
          >
             
          </p>
        </div>
        <h2
          className="MuiTypography-root-98 MuiTypography-body1-107 MuiTypography-colorTextPrimary-120"
          style={
            Object {
              "color": "#E53935",
            }
          }
        >
          
        </h2>
        <button
          className="MuiButtonBase-root-185 MuiButtonBase-disabled-186 MuiButton-root-159 MuiButton-text-161 MuiButton-outlined-167 MuiButton-sizeSmall-182 MuiButton-disabled-179"
          disabled={true}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "marginTop": 10,
            }
          }
          tabIndex="-1"
          type="submit"
        >
          <span
            className="MuiButton-label-160"
          >
            Change Password
          </span>
        </button>
      </form>
    </div>
  </div>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
  />
</div>
`;
